src/category_theory/limits/shapes/constructions/finite_products_of_binary_products.lean

```diff
@@ -0,0 +1,211 @@
+import category_theory.limits.shapes.binary_products
+import category_theory.limits.preserves.shapes.products
+import category_theory.limits.preserves.shapes.binary_products
+import category_theory.limits.shapes.pullbacks
+import category_theory.pempty
+import data.equiv.fin
+
+/-!
+# Constructing finite products from binary products and terminal.
+
+If a category has binary products and a terminal object then it has finite products.
+If a functor preserves binary products and the terminal object then it preserves finite products.
+
+# TODO
+
+Provide the dual results.
+Show the analogous results for functors which reflect or create (co)limits.
+-/
+
+universes v u u'
+
+noncomputable theory
+open category_theory category_theory.category category_theory.limits
+namespace category_theory
+
+variables {J : Type v} [small_category J]
+variables {C : Type u} [category.{v} C]
+variables {D : Type u'} [category.{v} D]
+
+/--
+Given `n+1` objects of `C`, a fan for the last `n` with point `c₁.X` and a binary fan on `c₁.X` and
+`f 0`, we can build a fan for all `n+1`.
+
+In `extend_fan_is_limit` we show that if the two given fans are limits, then this fan is also a
+limit.
+-/
+@[simps {rhs_md := semireducible}]
+def extend_fan {n : ℕ} {f : ulift (fin (n+1)) → C}
+  (c₁ : fan (λ (i : ulift (fin n)), f ⟨i.down.succ⟩))
+  (c₂ : binary_fan (f ⟨0⟩) c₁.X) :
+  fan f :=
+fan.mk c₂.X
+begin
+  rintro ⟨i⟩,
+  revert i,
+  refine fin.cases _ _,
+  { apply c₂.fst },
+  { intro i,
+    apply c₂.snd ≫ c₁.π.app (ulift.up i) },
+end
+
+/--
+Show that if the two given fans in `extend_fan` are limits, then the constructed fan is also a
+limit.
+-/
+def extend_fan_is_limit {n : ℕ} (f : ulift (fin (n+1)) → C)
+  {c₁ : fan (λ (i : ulift (fin n)), f ⟨i.down.succ⟩)} {c₂ : binary_fan (f ⟨0⟩) c₁.X}
+  (t₁ : is_limit c₁) (t₂ : is_limit c₂) :
+  is_limit (extend_fan c₁ c₂) :=
+{ lift := λ s,
+  begin
+    apply (binary_fan.is_limit.lift' t₂ (s.π.app ⟨0⟩) _).1,
+    apply t₁.lift ⟨_, discrete.nat_trans (λ i, s.π.app ⟨i.down.succ⟩)⟩
+  end,
+  fac' := λ s,
+  begin
+    rintro ⟨j⟩,
+    apply fin.induction_on j,
+    { apply (binary_fan.is_limit.lift' t₂ _ _).2.1 },
+    { rintro i -,
+      dsimp only [extend_fan_π_app],
+      rw [fin.cases_succ, ← assoc, (binary_fan.is_limit.lift' t₂ _ _).2.2, t₁.fac],
+      refl }
+  end,
+  uniq' := λ s m w,
+  begin
+    apply binary_fan.is_limit.hom_ext t₂,
+    { rw (binary_fan.is_limit.lift' t₂ _ _).2.1,
+      apply w ⟨0⟩ },
+    { rw (binary_fan.is_limit.lift' t₂ _ _).2.2,
+      apply t₁.uniq ⟨_, _⟩,
+      rintro ⟨j⟩,
+      rw assoc,
+      dsimp only [discrete.nat_trans_app],
+      rw ← w ⟨j.succ⟩,
+      dsimp only [extend_fan_π_app],
+      rw fin.cases_succ }
+  end }
+
+section
+variables [has_binary_products.{v} C] [has_terminal C]
+
+/--
+If `C` has a terminal object and binary products, then it has a product for objects indexed by
+`ulift (fin n)`.
+-/
+lemma has_product_ulift_fin :
```

Comment:
Do you want to mark some of these lemmas/defs as implementation detail? Either making the `private`, or flagging this in the comments?

Reply:
I did both :)